packages/vote/fpc/fpc.go

```diff
@@ -0,0 +1,267 @@
+	f.lastRoundCompletedSuccessfully = err == nil
+	return err
+}
+
+// enqueues items for voting
+func (f *FPC) enqueue() {
+	f.queueMu.Lock()
+	defer f.queueMu.Unlock()
+	f.ctxsMu.Lock()
+	defer f.ctxsMu.Unlock()
+	for ele := f.queue.Front(); ele != nil; ele = f.queue.Front() {
+		voteCtx := ele.Value.(*VoteContext)
+		f.ctxs[voteCtx.id] = voteCtx
+		f.queue.Remove(ele)
+		delete(f.queueSet, voteCtx.id)
+	}
+}
+
+// formOpinions updates the opinion for ongoing vote contexts by comparing their liked percentage
+// against the threshold appropriate for their given rounds.
+func (f *FPC) formOpinions(rand float64) {
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	for _, voteCtx := range f.ctxs {
+		// when the vote context is new there's no opinion to form
+		if voteCtx.IsNew() {
+			continue
+		}
+
+		lowerThreshold := f.paras.SubsequentRoundsLowerBoundThreshold
+		upperThreshold := 1 - f.paras.SubsequentRoundsLowerBoundThreshold
+
+		if voteCtx.HadFirstRound() {
+			lowerThreshold = f.paras.FirstRoundLowerBoundThreshold
+			upperThreshold = f.paras.FirstRoundUpperBoundThreshold
+		}
+
+		if voteCtx.Liked >= RandUniformThreshold(rand, lowerThreshold, upperThreshold) {
+			voteCtx.AddOpinion(vote.Like)
+			continue
+		}
+		voteCtx.AddOpinion(vote.Dislike)
+	}
+}
+
+// emits a Voted event for every finalized vote context (or Failed event if failed) and then removes it from FPC.
+func (f *FPC) finalizeOpinions() {
+	f.ctxsMu.Lock()
+	defer f.ctxsMu.Unlock()
+	for id, voteCtx := range f.ctxs {
+		if voteCtx.IsFinalized(f.paras.CoolingOffPeriod, f.paras.FinalizationThreshold) {
+			f.events.Finalized.Trigger(id, voteCtx.LastOpinion())
+			delete(f.ctxs, id)
+		}
+		if voteCtx.Rounds >= f.paras.MaxRoundsPerVoteContext {
+			f.events.Failed.Trigger(id, voteCtx.LastOpinion())
+			delete(f.ctxs, id)
+			continue
+		}
+	}
+}
+
+// queries the opinions of QuerySampleSize amount of OpinionGivers.
+func (f *FPC) queryOpinions() error {
+	ids := f.voteContextIDs()
+
+	// nothing to vote on
+	if len(ids) == 0 {
+		return nil
+	}
+
+	opinionGivers, err := f.opinionGiverFunc()
+	if err != nil {
+		return err
+	}
+
+	// nobody to query
+	if len(opinionGivers) == 0 {
+		return ErrNoOpinionGiversAvailable
+	}
+
+	// select random subset to query (the same opinion giver can occur multiple times)
+	opinionGiversToQuery := make([]vote.OpinionGiver, f.paras.QuerySampleSize)
+	for i := 0; i < f.paras.QuerySampleSize; i++ {
+		opinionGiversToQuery[i] = opinionGivers[f.rng.Intn(len(opinionGivers))]
+	}
+
+	// votes per id
+	var voteMapMu sync.Mutex
+	voteMap := map[string]vote.Opinions{}
+
+	// send queries
+	var wg sync.WaitGroup
+	for _, opinionGiverToQuery := range opinionGiversToQuery {
+		wg.Add(1)
+		go func(opinionGiverToQuery vote.OpinionGiver) {
+			defer wg.Done()
+
+			queryCtx, cancel := context.WithTimeout(context.Background(), f.paras.QueryTimeout)
+			defer cancel()
+
+			// query
+			opinions, err := opinionGiverToQuery.Query(queryCtx, ids)
```

Comment:
Maybe as a future improvment we might consider to not re-query the same `opinionGiverToQuery` (if sampled multiple times) multiple times and instead re-use its opinion. This should save some network resources.

Reply:
I've modified this now to be a set of OpinionGiver so the actual query is done only once.